### PR TITLE
Remove support for "provider" in cloud provider configs

### PR DESCRIPTION
**Pull request by rallytime**
The "driver" key should now be used instead of "provider", as the
"provider" key should only be used is cloud profile config files.

Deprecations warnings have been in place for this change since Salt
2015.5.0 came out and were slated for removal in the ``Nitrogen``
release.

The original deprecation was made in PR #24796

ping @techhat for review.

### CODE REVIEW

File: salt/cloud/__init__.py [hunk 1]
```diff
@@ -1183,11 +1171,6 @@ def create(self, vm_, local_master=True):
             'minion', vm_, self.opts, default={}
         )
 
-        # Since using "provider: <provider-engine>" is deprecated, alias provider
-        # to use driver: "driver: <provider-engine>"
-        if 'provider' in vm_:
-            vm_['driver'] = vm_.pop('provider')
-
```
Comment (eradman) [inline salt/cloud/__init__.py]: There is an issue here: `provider.pop('provider')` returns a result that can be split in two parts:

    > /home/eradman/local/saltenv/lib/python2.7/site-packages/salt/cloud/__init__.py(1192)create()
    -> alias, driver = vm_['driver'].split(':')
    (Pdb) vm_['driver']
    'rad_joyent_account:joyent'

Whereas after this change `vm_['driver']` only contained the string with the driver

    > /home/eradman/local/saltenv/lib/python2.7/site-packages/salt/cloud/__init__.py(1175)create()
    -> alias, driver = vm_['driver'].split(':')
    (Pdb) vm_['driver']
    'joyent'

**Review by eradman — CHANGES_REQUESTED** [review]
It will be great to see the ambiguous use of `provider` resolved!

File: salt/cloud/__init__.py (hunk 1 above)
Comment (rallytime) [inline salt/cloud/__init__.py]: @eradman Where are you setting the `rad_joyent_account:joyent` configuration? Perhaps I missed a use case when I originally made the deprecation, which of course would still be working if you hadn't moved over to `driver` yet.

**Comment by eradman** [comment]
I may be exploiting deprecated behavior. Here is my config:

_/etc/salt/cloud.providers.d/joyent.conf_

    rad_joyent_account:
      driver: joyent
      user: eradman
      password: **********
      private_key: /etc/salt/pki/cloud/joyent.pem
      keyname: joyent

_/etc/salt/cloud.profiles.d/joyent-profiles.conf_

    joyent_kvm_750:
      provider: rad_joyent_account
      size: k4-highcpu-kvm-750M
      image: centos-7
      location: us-east-1

**Comment by rallytime** [comment]
@eradman That is a completely correct way to define your cloud setup. It looks like the way I was supporting the deprecated behavior until we could remove this was covering up a bug in the initial deprecation. Thanks for pointing this out. I'll clean this up.

**Comment by rallytime** [comment]
@eradman It appears that the original deprecation code was a bit over zealous in the cloud.__init__.py `create` function. That area of the code _should_ be looking for the `provider` key, as it is parsing the profile config, not the provider config (which uses the `driver` key).

Thanks again for pointing this out.

**Comment by eradman** [comment]
I did notice one other issue with AWS:

    File "/home/eradman/local/saltenv/lib/python2.7/site-packages/salt/utils/aws.py", line 339, in _sig_key
      kDate = _sign(('AWS4' + key).encode('utf-8'), date_stamp)
    TypeError: cannot concatenate 'str' and 'dict' objects

Where contents of `key` is an empty dict.

My config:

_/etc/salt/cloud.providers.d/aws.conf_

    rad_aws_account:
      id: ****
      key: '*****'
      keyname: radman-aws
      private_key: /etc/salt/pki/cloud/aws.pem
      driver: ec2
      location: us-east-1

_/etc/salt/cloud.profiles.d/amazon-profiles.conf_

    aws-dev:
      provider: rad_aws_account
      image: ami-c75a10d0
      size: t2.micro
      ssh_username: root
      script_args: -D
      network_interfaces:
      - DeviceIndex: 0
        SubnetId: subnet-3bf94a72
        PrivateIpAddresses:
          - Primary: True
        AssociatePublicIpAddress: True
        SecurityGroupId:
          - sg-9f644fe5

**Comment by rallytime** [comment]
Thanks again for your testing @eradman. It is very much appreciated! I have fixed the EC2 TypeError as well as fixed a bug when trying to delete VMs with this change. I had tested this on DigitalOcean as well as Linode (and now EC2 as well) and those worked with this change. (Well, mostly. I ran into an error with the Linode driver, but it is also present on the `develop` branch and is not related to this particular change. I need to look into that separately.)

We have some nightly cloud provider tests that I know @Ch3LL has been working on, but they do not run on PRs. I am curious to see what other issues those tests might expose, since there were definitely some bugs in the original deprecation that were masked by providing support for `provider` along with `driver`. (I want to be able to catch those bugs before releasing, which is why I wanted to get this into develop to bake well before Nitrogen is ready.)

**Comment by eradman** [comment]
@rallytime, glad to this path being cleaned up a bit, testing is no trouble at all for me. Keep up the good work!

**Comment by eradman** [comment]
+1, no issues with vmware

**Comment by rallytime** [comment]
@cachedout This is ready for a final review.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@rallytime Are you comfortable with having this merged with the concern from @eradman still outstanding?

**Comment by eradman** [comment]
Every issue I encountered with this PR has been resolved. Good to go from my perspective!

**Comment by cachedout** [comment]
@eradman Thanks for the confirmation! I'll get this in.

**Comment by rallytime** [comment]
Thanks guys!

**Pull request merged**